opentracing-okhttp3/src/main/java/io/opentracing/contrib/okhttp3/TracingInterceptor.java

```diff
@@ -0,0 +1,123 @@
+ * parent {@link io.opentracing.SpanContext} to {@link Request.Builder#tag(Object)}.
+ *
+ * @author Pavol Loffay
+ */
+public class TracingInterceptor implements Interceptor {
+
+    private Tracer tracer;
+    private List<SpanDecorator> decorators;
+
+    /**
+     * Create tracing interceptor. Interceptor has to be added to {@link OkHttpClient.Builder#addInterceptor(Interceptor)}
+     * and {@link OkHttpClient.Builder#addNetworkInterceptor(Interceptor)}.
+     *
+     * @param tracer tracer
+     * @param decorators decorators
+     */
+    public TracingInterceptor(Tracer tracer, List<SpanDecorator> decorators) {
+        this.tracer = tracer;
+        this.decorators = new ArrayList<>(decorators);
+    }
+
+    /**
+     * Add tracing interceptors to client builder.
+     *
+     * @param okBuilder client builder
+     * @param tracer tracer
+     * @param decorators span decorators
+     * @return client builder with added tracing interceptor
+     */
+    public static OkHttpClient.Builder addTracing(OkHttpClient.Builder okBuilder,
+                                                  Tracer tracer, List<SpanDecorator> decorators) {
+
+        TracingInterceptor tracingInterceptor = new TracingInterceptor(tracer, decorators);
+        return okBuilder.addInterceptor(tracingInterceptor)
+                    .addNetworkInterceptor(tracingInterceptor);
+    }
+
+    @Override
+    public Response intercept(Chain chain) throws IOException {
+        Response response;
+
+        // application interceptor?
+        if (chain.connection() == null) {
+            Tracer.SpanBuilder spanBuilder = tracer.buildSpan(chain.request().method())
+                    .withTag(Tags.SPAN_KIND.getKey(), Tags.SPAN_KIND_CLIENT);
+
+            Object tag = chain.request().tag();
+            TagWrapper tagWrapper;
+            if (tag != null && tag instanceof TagWrapper) {
+                tagWrapper = (TagWrapper) tag;
+                if (tagWrapper.getParentSpanContext() != null) {
+                    spanBuilder.asChildOf(tagWrapper.getParentSpanContext());
+                }
+            } else {
+                tagWrapper = new TagWrapper(tag);
+            }
+
+            Span span = spanBuilder.start();
+
+            for (SpanDecorator spanDecorator: decorators) {
+                spanDecorator.onRequest(chain.request(), span);
+            }
+
+            Request.Builder requestBuilder = chain.request().newBuilder();
+            tracer.inject(span.context(), Format.Builtin.HTTP_HEADERS, new RequestBuilderInjectAdapter(requestBuilder));
+
+            requestBuilder.tag(new TagWrapper(tagWrapper, span));
+
+            try {
+                response = chain.proceed(requestBuilder.build());
+
+                for (SpanDecorator spanDecorator: decorators) {
+                    spanDecorator.onResponse(response, span);
+                }
+            } catch (Throwable ex) {
+                for (SpanDecorator spanDecorator: decorators) {
```

Comment:
not sure how opentracing works, but I think it’s weird that spans get onErrors for other spans